javaslang/src/main/java/javaslang/collection/Traversable.java

```diff
@@ -447,6 +447,21 @@ default T single() {
     <C> Map<C, ? extends Traversable<T>> groupBy(Function<? super T, ? extends C> classifier);
 
     /**
+     * Matches each element with a unique key that you extract from it.
+     * If the same key is present twice, the function will return {@code None}.
+     *
+     * @param classifier A function which extracts a key from elements
+     * @param <K>        key class type
+     * @return A Map containing the elements arranged by their keys.
+     * @throws NullPointerException if {@code getKey} is null.
+     */
+    default <K> Option<Map<K, T>> arrangeBy(Function<? super T, K> getKey) {
+        return Option.of(groupBy(getKey).mapValues(Traversable<T>::singleOption))
+            .filter(map -> !map.exists(kv -> kv._2.isEmpty()))
+            .map(map -> map.mapValues(Option::get));
+    }
```

Comment:
This is really a lovely method.

## Name

I like the name.

The method is similar to groupBy(). I first thought about naming it groupDistinctBy() but the 'group' part suggests that there may be several elements grouped together. That contradicts the 'distinct' part of the name. 'Arranging' elements is perfect!

## Signature

Because groupBy() and arrangeBy() are so similar I would suggest to 'sync' their signatures, i.e.

```java
<C> Map<C, ? extends Traversable<T>> groupBy(Function<? super T, ? extends C> classifier);

<C> Option<Map<C, T>> arrangeBy(Function<? super T, ? extends C> classifier);
```

In order to make the `? extends C` part work, we need to narrow the Map:

```java
default <C> Option<Map<C, T>> arrangeBy(Function<? super T, ? extends C> classifier) {
    return Option.of(groupBy(classifier).mapValues(Traversable::singleOption))
            .filter(map -> !map.exists(entry -> entry._2.isEmpty()))
            .map(map -> Map.narrow(map.mapValues(Option::get)));
}
```

**Question:** Can something bad happen when treating a `Map<? extends K, V>` as `Map<K, V>`? We know that persistent/immutable objects are allowed to be narrowed (see [this post](http://blog.javaslang.io/why-java-collections-framework-is-a-cats-and-dogs-library)).

But a Map heavily relies on equals/hashCode. If we deal with a type hierarchy, `a.equals(b)` might be true while `b.equals(a)` might be false, for `A a`, `B b`, where `B` extends `A`.

If the classifier type `Function<? super T, ? extends C>` should be `Function<? super T, C>`, we will have to adjust several other methods in Javaslang, starting with `groupBy(Function)`.

## Return type

I thought a minute about returning `Option<? extends Map<C, T>>` because a LinkedHashMap might want to preserve the insertion order of arranged elements. But we can't preserve the sorted property of a TreeMap because the classifier might not be comparable.

I think starting with `Option<Map<C, T>>` is the right way. It behaves like groupBy().